frame/dvm/evm/precompiles/mpt/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+	VerifyMultiStorageProof =
+		"verify_multi_storage_proof(bytes32,address,bytes[],bytes32[],bytes[][])",
+}
+
+const MAX_MULTI_STORAGEKEY_SIZE: usize = 32;
+const VERIFY_SINGLE_STORAGE_GAS: usize = 50_000;
+
+pub struct MPT<T> {
+	_marker: PhantomData<T>,
+}
+
+impl<T> Precompile for MPT<T>
+where
+	T: darwinia_evm::Config,
+{
+	fn execute(
+		input: &[u8],
+		target_gas: Option<u64>,
+		context: &Context,
+		is_static: bool,
+	) -> PrecompileResult {
+		let helper = PrecompileHelper::<T>::new(input, target_gas);
+		let (selector, data) = helper.split_input()?;
+		let action = Action::from_u32(selector)?;
+
+		// Check state modifiers
+		helper.check_state_modifier(context, is_static, StateMutability::View)?;
+
+		let mut multiplier: usize = 1;
+
+		let output = match action {
+			Action::VerfiySingleStorageProof => {
+				let params = MPTSingleStorageVerifyParams::decode(data)
+					.map_err(|_| helper.revert("decode single storage verify info failed"))?;
+				let proof = EthereumStorageProof::new(
+					params.lane_address,
+					params.storage_key,
+					params.account_proof,
+					params.storage_proof,
+				);
+				let storage_value =
+					EthereumStorage::<Vec<u8>>::verify_storage_proof(params.state_root, &proof)
+						.map_err(|_| helper.revert("verify single storage proof failed"))?;
+				abi_encode_bytes(storage_value.0.as_slice())
+			},
+			Action::VerifyMultiStorageProof => {
+				let params = MPTMultiStorageVerifyParams::decode(data)
+					.map_err(|_| helper.revert("decode multi storage verify info failed"))?;
+				let key_size = params.storage_keys.len();
+				if key_size != params.storage_proofs.len() {
+					return Err(helper.revert("storage keys not match storage proofs"));
+				}
+				if key_size > MAX_MULTI_STORAGEKEY_SIZE {
+					return Err(helper.revert("storage keys size too large"));
+				}
+
+				multiplier = key_size;
+
+				let storage_values: Result<Vec<Vec<u8>>, _> = (0..key_size)
+					.map(|idx| {
+						let storage_key = params.storage_keys[idx];
+						let storage_proof = params.storage_proofs[idx].clone();
+						let proof = EthereumStorageProof::new(
+							params.lane_address,
+							storage_key,
+							params.account_proof.clone(),
+							storage_proof,
+						);
+						let storage_value = EthereumStorage::<Vec<u8>>::verify_storage_proof(
+							params.state_root,
+							&proof,
+						);
+						match storage_value {
+							Ok(value) => {
+								return Ok(value.0.clone());
+							},
+							Err(err) => {
+								// if the key not exist, we should return Zero Value due to the
+								// Zero Value will not be stored.
+								if err == StorageProofError(ProofError::TrieKeyNotExist) {
+									return Ok(vec![]);
+								} else {
+									return Err(helper.revert("verfiy storage failed"));
+								}
+							},
+						}
+					})
+					.collect();
+				abi_encode_array_bytes(storage_values?)
+			},
+		};
+
+		let cost: u64 = u64::try_from(multiplier * VERIFY_SINGLE_STORAGE_GAS)
```

Comment:
this multiply may overflow for usize type. `checked_mul` is better.

Reply:
Nice catch, let me fix it.

Reply:
Tip: use `saturating_mul` to get the `MAX` if overflow.

Reply:
multiplier < MAX_MULTI_STORAGEKEY_SIZE, and overflow could not happend.

Reply:
remove MAX_MULTI_STORAGEKEY_SIZE and use `checked_mul` may be better.

Reply:
Why is `MAX_MULTI_STORAGEKEY_SIZE` designed previously? @hujw77 @xiaoch05

Reply:
I think `MAX_MULTI_STORAGEKEY_SIZE` is unnecessary.

Reply:
`MAX_MULTI_STORAGEKEY_SIZE ` is indeed unnecessary.
We should use gas cost to limit this size.

Reply:
done